mac/KeymanEngine4Mac/KeymanEngine4Mac/CoreWrapper/CoreWrapper.m

```diff
@@ -0,0 +1,395 @@
+          [self.coreHelper logDebugMessage:@"createCoreActionForActionStruct converted marker backspace, expected value =%lu, expected type =%u", backspace.expected_value, backspace.expected_type];
+        } else {
+          [self.coreHelper logDebugMessage:@"createCoreActionForActionStruct did not convert unknown backspace, expected value =%lu, expected type =%u", backspace.expected_value, backspace.expected_type];
+        }
+        break;
+      }
+      case KM_CORE_IT_PERSIST_OPT: {
+        [self.coreHelper logDebugMessage:@"***createCoreActionForActionStruct Persist Options encountered."];
+        km_core_option_item const * option = actionStruct->option;
+        NSString *keyString = [self.coreHelper createNSStringFromUnicharString:option->key];
+        NSString *valueString = [self.coreHelper createNSStringFromUnicharString:option->value];
+        
+        [self.coreHelper logDebugMessage:@"***createCoreActionForActionStruct converted Persist Options, key = %@, value = %@, scope = %d", keyString, valueString, option->scope];
+        
+        action = [[CoreAction alloc] initPersistOptionAction:keyString value:valueString scope:option->scope];
+        break;
+      }
+      case KM_CORE_IT_EMIT_KEYSTROKE: {
+        action = [[CoreAction alloc] initWithType: EmitKeystrokeAction actionContent:@"" backspaceCount:0 key:@"" value:@"" scope:0];
+        break;
+      }
+      case KM_CORE_IT_INVALIDATE_CONTEXT: {
+        action = [[CoreAction alloc] initWithType: InvalidateContextAction actionContent:@"" backspaceCount:0 key:@"" value:@"" scope:0];
+        break;
+      }
+      case KM_CORE_IT_CAPSLOCK: {
+        action = [[CoreAction alloc] initWithType: CapsLockAction actionContent:@"" backspaceCount:0 key:@"" value:@"" scope:0];
+        break;
+      }
+      default: {
+        NSLog(@"createCoreActionForActionStruct unrecognized type of km_core_action_item = %u\n", actionStruct->type);
+      }
+  }
+  return action;
+}
+
+-(NSString *)getContextAsStringUsingCore {
+  km_core_context * context =  km_core_state_context(self.coreState);
+  
+  km_core_context_item * contextItemsArray = nil;
+  size_t contextLength = km_core_context_length(context);
+  
+  NSMutableString *contextString = [[NSMutableString alloc]init];
+
+  if (contextLength==0) {
+    [self.coreHelper logDebugMessage:@"***context is empty."];
+  } else {
+    km_core_status result = km_core_context_get(context, &contextItemsArray);
+    if (result==KM_CORE_STATUS_OK) {
+      for (int i = 0; i < contextLength; i++) {
+        km_core_context_item contextItem = contextItemsArray[i];
+        if (contextItem.type == KM_CORE_CT_CHAR) {
+          const unichar unicodeChar = contextItem.character;
+          NSString *charString = [NSString stringWithCharacters:&unicodeChar length:1];
+          [contextString appendString:charString];
+        }
+      }
+    }
+  }
```

Comment:
Is this correct for surrogate pairs (i.e. >= U+10000)?

Reply:
`unichar` says it's UTF-16 so this wouldn't be correct.   Line 277 is a narrowing of the type from 32 to 16 bits. 

in C++ i have the  `Utf32CharToUtf16()` macro to convert utf-32 to a 1-or-2-code unit string

Reply:
can you use https://developer.apple.com/documentation/swift/unicode/scalar instead ?

Reply:
Thanks for catching this! I had created a helper function for this conversion but neglected to use it here. I've added a unit test to ensure this works for a context string containing emojis.